## String Chains 

Given an array of words representing your dictionary, you test each word to see if it can be made into another word in the dictionary. This will be done by removing characters one at a time. Eac word represents its own first element of its string chain, so start with a string chain length of 1. Each time you remove a character, increment your string chain by 1. In order to remove a character, the resulting word must be in your original dictionary. Your goal is to determine the longest string chain achievable for a given dictionary. 

For example, a dictionary [a, and, an ,bear], the word and could be reduced to an and then to a. The single character a cannot be reduced any further as the null string is not in the dictionary. This would be the longest string chain, having a length 3. The word bear cannot be reduced at all. 

**Function Description**
Complete the function longestChain in the editor below. The function must return a single integer representing the length of the longest string chain. 

## Longest Word in Dictionary 

Given a list of strings words representing an English Dictionary, find the longest word in words that can be built one character at a time by other words in words. If there is more that one possible answer, return the longest word with the smallest lexicographical order. 

If there is not answer, return the empty string. 

**Example 1**:
Input:
words = ["w", "wo", "wor", "worl", "world"]
Output: "world"
Explanation:
The word "world" can be built one character at a time by "w", "wo", "wor", and "worl". 

**Example 2**:
Input:
words = ["a", "banana", "app", "appl", "ap", "apply", "apple"]
Output: "apple"
Explanation:
Both "apply" and "apple" can be built from other words in the dictionary. However, "apple" is lexicographlically smaller than "apply". 

### Approach 1: Brute Force 

**Intuition**:
For each word, check if all prefixes word[:k] are present. We can use a Set structure to check this quickly. 

**Algorithm**:
Whenever our found word would be superior, we check if all it's prefixes are present, then replace our answer. 

Alternatively, we could have sorted the words beforehand, so that we know the word we are considering would be the answer if all it's prefixes are present. 

In [1]:
def longestWord(words):
    ans = ""
    wordset = set(words)
    for word in words:
        if len(word) > len(ans) or len(word) == len(ans) and word < ans:
            if all(word[:k] in wordset for k in range(1,len(word))):
                ans = word
    return ans 

In [2]:
words = ["w", "wo", "wor", "worl", "world"]

longestWord(words)

'world'

In [3]:
"apple" < "apply"

True

In [4]:
words = ["a", "ap", "apple", "apply", "app", "appl"]

In [5]:
longestWord(words)

'apple'

In [7]:
words = ["a", "ap","apply", "app", "appl"]

In [8]:
longestWord(words)

'apply'

**Complexity Analysis**:

* Time complexity: $O(\sum w_i^2)$, where $w_i$ is the length of words[i]. Checking whether all prefixes of words[i] are in the set is $O(\sum w_i^2)$. 

* Space complexity: $O(\sum w_i^2)$. 

### Approach 2: Trie + Depth-First Search 

**Intuition**:
As prefixes of strings are involved, this is usually a natural fit for a trie (a prefix tree.)

**Algorithm**:
Put every word in a trie, then depth-first-search from the start of the trie, only searching nodes that ended a word. Every node found (except the root, which is a special case) then represents a word with all it's prefixes present. We take the best such word. 

In Python, we showcase a method using defaultdict, while in Java, we stick to a more general object-oriented approach. 

In [17]:
import collections
from functools import reduce 

In [20]:
def longestWord_2(words):
    Trie = lambda: collections.defaultdict(Trie)
    trie = Trie()
    END = True
    
    for i, word in enumerate(words):
        reduce(dict.__getitem__, word, trie)[END] = i
    
    stack = trie.values()
    ans = ""
    while stack:
        cur = stack.pop()
        if END in cur:
            word = words[cur[END]]
            if len(word) > len(ans) or len(word) == len(ans) and word < ans:
                ans = word
            stack.extend([cur[letter] for letter in cur if letter != END])
            
    return ans 

In [21]:
longestWord_2(words)

AttributeError: 'dict_values' object has no attribute 'pop'

In [22]:
# examples of default_dict 

s = [('yellow',1), ('blue', 2), ('yellow', 3), ('blue', 4), ('red',1)]

In [24]:
from collections import defaultdict
d = defaultdict(list)

In [25]:
for k,v in s:
    d[k].append(v)

In [26]:
d

defaultdict(list, {'yellow': [1, 3], 'blue': [2, 4], 'red': [1]})

When each key is encountered for the first time, it is not already in the mapping; so an entry is automaticaly created using the default_factory function which returns an empty list. The list.append() operation then attaches the value to the new list. When keys are encountered again, the look-up proceeds normally (returning the list for that key) and the list.append() operation. adds anohter value to the list. This technique is simpler and faster than an equivalent technique using dict.setdefault():

In [27]:
d = {}
for k,v in s :
    d.setdefault(k, []).append(v)

In [28]:
d

{'yellow': [1, 3], 'blue': [2, 4], 'red': [1]}

In [29]:
d.items()

dict_items([('yellow', [1, 3]), ('blue', [2, 4]), ('red', [1])])

Setting the default_factory to int makes the defaultdict useful for counting (like a bag or multiset in other languages):

In [30]:
s = 'mississippi'
d = defaultdict(int)
for k in s:
    d[k] += 1 

In [31]:
d.items()

dict_items([('m', 1), ('i', 4), ('s', 4), ('p', 2)])

When a letter is first encountered, it is missing from the mapping, so the default_factory function calls int() to supply count of zero. The increment operation then builds up the count for each letter. 

The function int() which always returns zero is just a special case of constant functions. A faster and more flexible way to create constant functions is to use lambda function which can supply any constant value (not just zero):

In [33]:
def constant_factory(value):
    return lambda: value 

In [34]:
constant_factory(5)

<function __main__.constant_factory.<locals>.<lambda>()>

In [35]:
d = defaultdict(constant_factory('<Wuhan>'))
d.update(name = 'XiaoWang', action = 'ran')
print("%(name)s %(action)s to %(object)s" %d)

XiaoWang ran to <Wuhan>


Setting the default_factory to set makes the defaultdict useful for building a dictionary of sets:

In [36]:
s = [('red',1), ('blue',2), ('red',3), ('blue', 4), ('red',1), ('blue', 4)]
d = defaultdict(set)

for k,v in s:
    d[k].add(v)

In [37]:
d.items()

dict_items([('red', {1, 3}), ('blue', {2, 4})])

## Valid IP address 

Write a function to check whether an input string is a valid IPv4 address or IPv6 address or neither.

IPv4 addresses are canonically represented in dot-decimal notation, which consists of four decimal numbers, each ranging from 0 to 255, separated by dots ("."), e.g.,172.16.254.1;

Besides, leading zeros in the IPv4 is invalid. For example, the address 172.16.254.01 is invalid.

IPv6 addresses are represented as eight groups of four hexadecimal digits, each group representing 16 bits. The groups are separated by colons (":"). For example, the address 2001:0db8:85a3:0000:0000:8a2e:0370:7334 is a valid one. Also, we could omit some leading zeros among four hexadecimal digits and some low-case characters in the address to upper-case ones, so 2001:db8:85a3:0:0:8A2E:0370:7334 is also a valid IPv6 address(Omit leading zeros and using upper cases).

However, we don't replace a consecutive group of zero value with a single empty group using two consecutive colons (::) to pursue simplicity. For example, 2001:0db8:85a3::8A2E:0370:7334 is an invalid IPv6 address.

Besides, extra leading zeros in the IPv6 is also invalid. For example, the address 02001:0db8:85a3:0000:0000:8a2e:0370:7334 is invalid.

Note: You may assume there is no extra space or special characters in the input string.

Example 1:

Input: "172.16.254.1"

Output: "IPv4"

Explanation: This is a valid IPv4 address, return "IPv4".
Example 2:

Input: "2001:0db8:85a3:0:0:8A2E:0370:7334"

Output: "IPv6"

Explanation: This is a valid IPv6 address, return "IPv6".
Example 3:

Input: "256.256.256.256"

Output: "Neither"

Explanation: This is neither a IPv4 address nor a IPv6 address.

In [1]:
# Divide and Conquer 

def isvalid_IPv4(IP:str) -> str:
    nums = IP.split('.')
    print(nums)
    i = 1
    for x in nums:
        print(x)
        if len(x) == 0 or len(x) > 3:
            print('stop at clause 1'+ str(i))
            return "Neither"
        if (x[0] == '0' and len(x) != 1) or not x.isdigit() or int(x) > 255:
            print('stop at clause 2' + str(i))
            return "Neither"
        i += 1
            
    return "IPv4"

def isvalid_IPv6(IP:str) -> str:
    nums = IP.split(':')
    for x in nums:
        if len(x) == 0 or len(x) > 4 or not all(c in hexdigits for c in x):
            return "Neither"
    return "IPv6"


def validIPAddress(IP:str) -> str:
    if IP.count('.') == 3:
        return isvalid_IPv4(IP)
    elif IP.count(":") == 7:
        return isvalid_IPv6(IP)
    else:
        return "Neither"

In [2]:
IP = "172.16.254.1"
validIPAddress(IP)

['172', '16', '254', '1']
172
16
254
1


'IPv4'

In [11]:
x = '254'

In [12]:
(x[0] == '0' and len(x) != 1)

False

In [13]:
not x.isdigit() 

False

In [3]:
IP = '1.0.1.'

validIPAddress(IP)

['1', '0', '1', '']
1
0
1

stop at clause 14


'Neither'

In [6]:
dp = [[0]*5]*5

In [7]:
print(dp)

[[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]


In [8]:
dp[0]

[0, 0, 0, 0, 0]

In [9]:
matrix = [[1,0,0,1,1], [1,1,1,1,1], [0,1,1,0,1], [0,0,0,1,1], [0,0,0,0,1]]

In [10]:
dp[0] = matrix[0]

In [11]:
dp[0]

[1, 0, 0, 1, 1]

In [12]:
dp[:][0] = matrix[:][0]

In [13]:
dp

[[1, 0, 0, 1, 1],
 [0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0]]

In [14]:
matrix

[[1, 0, 0, 1, 1],
 [1, 1, 1, 1, 1],
 [0, 1, 1, 0, 1],
 [0, 0, 0, 1, 1],
 [0, 0, 0, 0, 1]]

In [15]:
matrix[:][0]

[1, 0, 0, 1, 1]

In [16]:
matrix[0]

[1, 0, 0, 1, 1]

In [20]:
for j in range(5,1,-1):
    print(j)

5
4
3
2


In [24]:
dp = [ [matrix[i][j] for i in range(5)] for j in range(5)]

In [25]:
dp

[[1, 1, 0, 0, 0],
 [0, 1, 1, 0, 0],
 [0, 1, 1, 0, 0],
 [1, 1, 0, 1, 0],
 [1, 1, 1, 1, 1]]

In [26]:
matrix = [[i*j for i in range(1,6)] for j in range(1,4)]

In [27]:
matrix 

[[1, 2, 3, 4, 5], [2, 4, 6, 8, 10], [3, 6, 9, 12, 15]]

In [28]:
n = len(matrix)

m = len(matrix[0])

In [29]:
n

3

In [30]:
m

5

In [68]:
dp = [[0]*m]*n

In [69]:
dp

[[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]

In [70]:
dp[0][0:m] = matrix[0][0:m]

In [71]:
dp

[[1, 2, 3, 4, 5], [1, 2, 3, 4, 5], [1, 2, 3, 4, 5]]

In [72]:
dp[0:n][0] = matrix[0:n][0]

In [73]:
dp

[[1, 2, 3, 4, 5], [1, 2, 3, 4, 5], [1, 2, 3, 4, 5]]

In [41]:
matrix[0][0]

1

In [42]:
matrix[0][1]

2

In [43]:
matrix[0][2]

3

In [44]:
matrix[0][3]

4

In [75]:
int('1') - int('0')

1

In [135]:
matrix = [ ['1','0','1','0','0'],
           ['1', '0', '1', '1', '1'],
           ['1', '1', '1', '1', '1'],
           ['1', '0', '0', '1', '0'] ]

In [136]:
n = len(matrix)
m = len(matrix[0])
n
m

5

In [137]:
dp = [[int(matrix[j][i]) for i in range(m)] for j in range(n)]

In [138]:
dp

[[1, 0, 1, 0, 0], [1, 0, 1, 1, 1], [1, 1, 1, 1, 1], [1, 0, 0, 1, 0]]

In [139]:
for j in range(1,n):
    dp[j][0] = dp[j-1][0] + int(matrix[j][0])
    
for i in range(1,m):
    dp[0][i] = dp[0][i-1] + int(matrix[0][i])

In [140]:
dp

[[1, 1, 2, 2, 2], [2, 0, 1, 1, 1], [3, 1, 1, 1, 1], [4, 0, 0, 1, 0]]

In [141]:
for j in range(1,n):
    for i in range(1,m):
        dp[j][i] = dp[j-1][i] + dp[j][i-1] - dp[j-1][i-1] + int(matrix[j][i])

In [142]:
dp

[[1, 1, 2, 2, 2], [2, 2, 4, 5, 6], [3, 4, 7, 9, 11], [4, 5, 8, 11, 13]]

In [134]:
matrix

[['1', '0', '1', '0', '0'],
 ['1', '0', '1', '1', '1'],
 ['1', '1', '1', '1', '1'],
 ['1', '0', '0', '1', '0']]

In [143]:
ans = 0
for j in range(1,n):
    for i in range(1,m):
        for size in range(min(n-j, m-i)-1,0,-1):
            temp = dp[j + size -1][ i+ size -1] - dp[j-1][i+size -1] - dp[j + size -1][i -1] + dp[j-1][i-1]
            if temp == size*size:
                ans = max(ans, temp)

In [144]:
ans

4

In [168]:
def maximalSquare( matrix: list) -> int:
    ans = 0


    n = len(matrix)
    m = len(matrix[0])
    
    print('n is '+ str(n) + ' and m is '+ str(m))

    if n == 0:
        print('n is 0')
        return ans
    

    dp = [[int(matrix[j][i]) for i in range(m)] for j in range(n)]
    
    print(dp)


    for j in range(1,n):
         dp[j][0] = dp[j-1][0] + int(matrix[j][0])

    for i in range(1,m):
         dp[0][i] = dp[0][i-1] + int(matrix[0][i])

    for j in range(1,n):
        for i in range(1,m):
            dp[j][i] = dp[j-1][i] + dp[j][i-1] - dp[j-1][i-1] + int(matrix[j][i])
    
    print(dp)
    
    for j in range(1,n):
        for i in range(1,m):
            for size in range(min(n-j, m-i)-1,0,-1):
                temp = dp[j + size -1][ i+ size -1] - dp[j-1][i+size -1] - dp[j + size -1][i -1] + dp[j-1][i-1]
                if temp == size*size:
                    print(size)
                    print(temp)
                    ans = max(ans, temp)

    return ans 

In [157]:
maximalSquare(matrix)

[[1, 0, 1, 0, 0], [1, 0, 1, 1, 1], [1, 1, 1, 1, 1], [1, 0, 0, 1, 0]]
[[1, 1, 2, 2, 2], [2, 2, 4, 5, 6], [3, 4, 7, 9, 11], [4, 5, 8, 11, 13]]
2
4
1
1
1
1
1
1
1
1
1
1


4

In [17]:
matrix = [["0","1"],["1","0"]]

In [169]:
maximalSquare(matrix)

n is 2 and m is 2
[[0, 1], [1, 0]]
[[0, 1], [1, 2]]


0

In [15]:
def maximalSquare2(matrix : list) -> str:
    if not matrix:
        return 0
    n = len(matrix)
    m = len(matrix[0])
    
    sizes = [[0 for i in range(m)] for j in range(n)]
    print(sizes)
    ans = 0
    
    sizes[0][0] = int(matrix[0][0])
    
    for j in range(m):
        for i in range(n):

            if int(matrix[i][j])  == 1:
            
                if (i==0 and j >= 1):
                   # sizes[i][j] = sizes[i][j-1] + 1
                    sizes[i][j] = int(matrix[i][j])
                elif (j==0 and i >= 1):
                   # sizes[i][j] = sizes[i-1][j] + 1
                    sizes[i][j] = int(matrix[i][j])
                elif (j >= 1 and i >= 1):
                    sizes[i][j] = min(min(sizes[i-1][j-1], sizes[i-1][j]), sizes[i][j-1]) +1
                
            ans = max(ans, sizes[i][j]*sizes[i][j])
            
            print(sizes)
    return ans
                

In [18]:
maximalSquare2(matrix)

[[0, 0], [0, 0]]
[[0, 0], [0, 0]]
[[0, 0], [1, 0]]
[[0, 1], [1, 0]]
[[0, 1], [1, 0]]


1

In [19]:
matrix = [ ['1','0','1','0','0'],
           ['1', '0', '1', '1', '1'],
           ['1', '1', '1', '1', '1'],
           ['1', '0', '0', '1', '0'] ]

In [20]:
maximalSquare2(matrix)

[[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]
[[1, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]
[[1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]
[[1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [0, 0, 0, 0, 0]]
[[1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [1, 0, 0, 0, 0]]
[[1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [1, 0, 0, 0, 0]]
[[1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [1, 0, 0, 0, 0]]
[[1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [1, 1, 0, 0, 0], [1, 0, 0, 0, 0]]
[[1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [1, 1, 0, 0, 0], [1, 0, 0, 0, 0]]
[[1, 0, 1, 0, 0], [1, 0, 0, 0, 0], [1, 1, 0, 0, 0], [1, 0, 0, 0, 0]]
[[1, 0, 1, 0, 0], [1, 0, 1, 0, 0], [1, 1, 0, 0, 0], [1, 0, 0, 0, 0]]
[[1, 0, 1, 0, 0], [1, 0, 1, 0, 0], [1, 1, 1, 0, 0], [1, 0, 0, 0, 0]]
[[1, 0, 1, 0, 0], [1, 0, 1, 0, 0], [1, 1, 1, 0, 0], [1, 0, 0, 0, 0]]
[[1, 0, 1, 0, 0], [1, 0, 1, 0, 0], [1, 1, 1, 0, 0], [1, 0, 0, 0, 0]]
[[1, 0, 1, 0, 0], [1, 0, 1, 1, 0],

4

In [21]:
matrix = [['0','0','0'], ['0', '0', '0'], ['1', '1', '1']]

In [22]:
maximalSquare2(matrix)

[[0, 0, 0], [0, 0, 0], [0, 0, 0]]
[[0, 0, 0], [0, 0, 0], [0, 0, 0]]
[[0, 0, 0], [0, 0, 0], [0, 0, 0]]
[[0, 0, 0], [0, 0, 0], [1, 0, 0]]
[[0, 0, 0], [0, 0, 0], [1, 0, 0]]
[[0, 0, 0], [0, 0, 0], [1, 0, 0]]
[[0, 0, 0], [0, 0, 0], [1, 1, 0]]
[[0, 0, 0], [0, 0, 0], [1, 1, 0]]
[[0, 0, 0], [0, 0, 0], [1, 1, 0]]
[[0, 0, 0], [0, 0, 0], [1, 1, 1]]


1

## Python | Using 2D arrays/lists the right way

In [4]:
# Method 2a

rows, cols = (5,5)
arr = [[0] * cols] * rows 
print(arr)

[[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]


In [5]:
arr[0][0] = 1

print(arr)

[[1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [1, 0, 0, 0, 0], [1, 0, 0, 0, 0]]


In [2]:
# Method 2b 

rows, cols = (5,5)

arr = [[0 for i in range(cols)] for j in range(rows)]

print(arr)

[[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]


In [3]:
arr[0][0] = 1

print(arr)

[[1, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]


## 301 Remove Invalid Parentheses 

Remove the minimum number of invalid parentheses in order to make the input string valid. Return all possible results. 

Note: the input string may contain letters other than the parentheses '(' and ')'. 

Example 1:

Input: "()())()"
Output: ["()()()", "(())()"]
Example 2:

Input: "(a)())()"
Output: ["(a)()()", "(a())()"]
Example 3:

Input: ")("
Output: [""]

In [72]:
def dfs(s: str, start: int, l: int, r: int, ans: list):
    if not s:
        ans.append("")
        return 
    if l== 0 and r == 0 and isvalid(s):
        ans.append(s)
        return
    else:
        for i in range(start, len(s)):
            if i != start and i < len(s) and s[i] == s[i-1]:
                continue
            if s[i] == ')' and r > 0:
                print('i is '+ str(i))
                new_s = s[:i]+s[i+1:]
                print(new_s,l,r-1)
                dfs(new_s,i, l, r-1, ans)
            if s[i] == '(' and l > 0:
                print('i is '+ str(i))
                new_s = s[:i] + s[i+1:]
                print(new_s)
                dfs(new_s,i, l-1, r, ans)


def isvalid(s:str) -> bool:
    count = 0
    for ch in s:
        if ch == '(':
            count += 1
        elif ch == ')':
            count -= 1
        if count < 0:
            return False 
    return count == 0




def removeInvalidParentheses(s: str) -> list:
    
    if not s:
        return [""]
    elif '(' not in s and ')' not in s:
        return [s]
    elif isvalid(s):
        return [s]
    
    # count the number of left parenthese and right parentheses
    # to be deleted 
    l = 0
    r = 0
    
    for ch in s:
        if ch == '(':
            l += 1
        if ch == ')':
            if l == 0:
                r += 1
            elif l > 0:
                l -= 1
    ans = []
    dfs(s, 0, l, r, ans)
    return ans 


    
        

In [73]:
s = ")("

In [74]:
removeInvalidParentheses(s)

i is 0
( 1 0
i is 0

i is 1
)


['']

In [75]:
s = "()())()"

In [76]:
removeInvalidParentheses(s)

i is 1
(())() 0 0
i is 3
()()() 0 0
i is 6
()())( 0 0


['(())()', '()()()']

In [36]:
isvalid(s)

False

In [77]:
s = "(a)())()"

In [78]:
removeInvalidParentheses(s)

i is 2
(a())() 0 0
i is 4
(a)()() 0 0
i is 7
(a)())( 0 0


['(a())()', '(a)()()']

In [51]:
import collections 
class Solution:
    def removeInvalidParentheses(self, s: str) -> List[str]:
        
        if not s:
            return [""]
        elif '('  not in s and ')'  not in s:
            return [s]
        elif self.isvalid(s):
            return [s]
        
        queue = collections.deque([s])
        result = []
        visited = set()
        
        while queue:
            for i in range(len(queue)):
                curr_s = queue.popleft()
                if self.isvalid(curr_s):
                    result.append(curr_s)
                else:
                    for k in range(len(curr_s)):
                        temp = curr_s[0:k] + curr_s[k+1:]
                        if temp not in visited: 
                            visited.add(temp)
                            queue.append(temp)
            if result:
                return result
        else:
            return [""]
        
    def isvalid(self, s: str):
        count = 0
        for l in s:
            if l == '(':
                count += 1
            elif l == ')':
                count -= 1
            if count < 0:
                return False
        return count == 0
        
        

NameError: name 'List' is not defined

## 70. Climbing Stairs 

You are climbing a stair case. It takes n steps t reach to the top. 

Each time you can either climb 1 or 2 steps. In how may distinct ways can you climb to the top? 

**Note:** Given n will be a positive integer. 

**Example 1:** 
Input: 2
Output: 2 
Explanation: There are two ways to climb to the top.
1. 1 step + 1 step
2. 2 step

**Example 2:**
Input: 3
Output: 3
Explanation: There are three ways to climb to the top. 
1. 1 step + 1 step + 1 step
2. 1 step + 2 step 
3. 2 step + 1 step

**Solution:**
f(n) = f(n-1) + f(n-2) 
with f(1) = 1 and f(0) = 1

In [24]:
def climbStairs(n: int) -> int:
    if n <= 1:
        return 1
    else:
        a = 1
        b = 1
        for i in range(2,n+1):
            a, b = a+b, a
        return a

In [25]:
climbStairs(2)

2

## 62. Unique Paths

A robot is located at the top-left corner of a m*n grid(marked 'Start in the diagram below). 

The robot can only move either down or right at any point in time, The robot is trying to reach the bottom-right corner of the grid (marked 'Finish' in the diagram below). 

How many possible unique paths are there?

In [26]:
n = 3
m = 5
dp = [[0] *(n+1) for _ in range(m+1)] 

In [27]:
dp

[[0, 0, 0, 0],
 [0, 0, 0, 0],
 [0, 0, 0, 0],
 [0, 0, 0, 0],
 [0, 0, 0, 0],
 [0, 0, 0, 0]]

In [29]:
dp[4][3]

0

In [30]:
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        
        dp = [[0] *(n+1) for _ in range(m+1)] 
        
        for i in range(1,m+1):
            for j in range(1,n+1):
                if i == 1 and j == 1:
                    dp[i][j] = 1
                else:
                    dp[i][j] = dp[i-1][j] + dp[i][j-1]
        return dp[i][j]
        

Time complexity O(mn). Space complexity O(mn). 

## 926. Flip String to Monotone Increasing 

A string of '0's and '1's is monotone increasing if it consists of some number of '0's (possibly 0), followed by some number of 1's (also possibly 0.)

We are given a string s of '0' and '1's, and we may flip any '0' to a '1' or a '1' to a '0'.

Return the minimum number of flips to make s monotone increasing. 


**Example 1:**
Input: "00110"
Output: 1
Explanation: We flip the last digit to get 00111. 

**Example 2:**
Input: "010110"
Output: 2
Explanation: We flip to get 011111, or alternatively 000111. 

**Example 3:**
Input: "00011000"
Output: 2
Explanation: We flip to get 0000000. 

In [34]:
ord('1') - ord('0')

1

In [68]:
def miniFlipsMonoIncre(s: str) -> int:
    
    n = len(s)
    
    l =  [0]*(n+1)
    
    r = [0]*(n+1)
    
    
    for i in range(1,n+1):
        l[i] = l[i-1] + ord(s[i-1]) - ord('0')
        
    print(l)
        
    for i in range(n-1,-1,-1):
        r[i] = r[i+1] + ord('1') - ord(s[i]) 
        
    print(r)
    
    ans = r[0]    
    
    for i in range(1,n+1):
        ans = min(ans, l[i-1]+ r[i])
    return ans 
        
        

In [64]:
s = "00110"
miniFlipsMonoIncre(s)

[0, 0, 0, 1, 2, 2]
[3, 2, 1, 1, 1, 0]


1

In [65]:
s = "010110"
miniFlipsMonoIncre(s)

[0, 0, 1, 1, 2, 3, 3]
[3, 2, 2, 1, 1, 1, 0]


2

In [69]:
s = "00011000"
miniFlipsMonoIncre(s)

[0, 0, 0, 0, 1, 2, 2, 2, 2]
[6, 5, 4, 3, 3, 3, 2, 1, 0]


2

Time Complexity: O(n)

Space Complexity: O(n)

## 790. Domino and Tromino Tiling 

We have two types of tiles: a 2*1 domino shape, and an "L" tromino shape. These shapes may be rotated.

XX <- domino

XX <- "L" tromino 
X

Given N, how many ways are there to tile a 2*N board? Return your answer modulo 10^9 +7. 

**Example:**
Input: 3
Output: 5
Explanation: 
The five different ways are listed below, different letters indicates different tiles:

XYZ  XXZ  XXY   XYY   XYY

XYZ  YYZ  XYY   XZZ   XXY

In [232]:
def numTilings(N:int) -> int:
    
    dp =  [[0]*2 for i in range (N+1)]
    
    dp[0][0] = 1
    dp[1][0] = 1
    dp[0][1] = 1 
    
    for i in range(2, N+1):
        dp[i][0] = (dp[i-1][0] + dp[i-2][0] + 2*dp[i-1][1]) % (10**9+7)
        dp[i][1] = (dp[i-2][0] + dp[i-1][1])% (10**9 + 7)
    
    return (dp[N][0]% (10**9+7) )
        
    
    

Time: O(N) Space : O(N)

In [233]:
N = 3

numTilings(3)

5

In [230]:
5 % 3

2

In [244]:
def numTilings2(N:int) -> int:
    
    a = 1 
    b = 1
    c = 1
    d = 0 
    
    for i in range(2, N+1):
        

        
        a, b, d = b% (10**9 + 7), ( a + b + 2*d )% (10**9 + 7), (a+d)% (10**9 + 7)
        
        
    return (b %(10**9 + 7))
        
    

In [245]:
numTilings2(3)

5

## 818. Race Car

You car starts at position 0 and speed +1 on an infinite number line. (Your car can go inoto negative positions). 

You car dirves automatically accoding to a sequence of instructions A (accelerate) and R (reverse). 

When you get an instruction "A", your car does the following: position += speed, speed *= 2. 

When you get an instruction "R", your car does the following: if your speed is positive then speed = -1, otherwise speed = 1. (Your position stays the same.)

For example, after commands "AAR", your car goes to positions 0->1->3->3, and your speed goes to 1->2->4->-1. 

Now for some target position, say the length of the shortest sequence of instructions to get there. 

**Example 1:**
Input: target = 3
Output: 2
Explanation: 
The shortest instruction sequence is "AA". Your position goes from 0->1->3. 

**Example 2:**
Input: target = 6
Output: 5
Explanation:
The shortest instruction sequence is "AAARA". Your position goes from 0->1->3->7->7->6. 

In [127]:
import math 
math.ceil(2.5)

3

In [129]:
2**5

32

In [146]:
n = 5

for m in range(0,5):
    print(m)

0
1
2
3
4


In [224]:
# Recursion with memorization (DP)

def racecar(target: int) -> int:
    
    m_ = [0] * (target+1) 
    
    
    
    def racecar_dp(t: int):
        if t == 0 or t == 1:
            m_[t] = t
            return m_[t]
        
        if m_[t] > 0: 
            return m_[t]
        
        print('t is '+ str(t))
#         n = math.ceil(math.log(t,2))
        n = t.bit_length()
        print(n)
        if 2**n == t + 1:
            m_[t] = n 
            return m_[t]
        else:
            m_[t] = n + 1 + racecar_dp(2**n - 1 - t)
            print('remain is '+ str(2**n - 1 - t))
            print('step is '+ str(m_[t]))
            for m in range(0,n-1):
                cur = 2**(n-1) - 2**m
                print('curr is '+ str(cur))
                print('step is '+ str(n+m+1 + racecar_dp(t - cur)))
                m_[t] = min(m_[t], n+ m + 1 + racecar_dp(t - cur))
                print('smaller step is '+ str(m_[t]))
        return m_[t]
            
    racecar_dp(target)
    return m_
        

In [225]:
target = 3

racecar(target)

t is 3
2


[0, 0, 0, 2]

In [226]:
target = 4

racecar(target)

t is 4
3
t is 3
2
remain is 3
step is 6
curr is 3
step is 5
smaller step is 5
curr is 2
t is 2
2
remain is 1
step is 4
curr is 1
step is 4
smaller step is 4
step is 9
smaller step is 5


[0, 1, 4, 2, 5]

In [227]:
target = 6

racecar(target)

t is 6
3
remain is 1
step is 5
curr is 3
t is 3
2
step is 6
smaller step is 5
curr is 2
t is 4
3
remain is 3
step is 6
curr is 3
step is 5
smaller step is 5
curr is 2
t is 2
2
remain is 1
step is 4
curr is 1
step is 4
smaller step is 4
step is 9
smaller step is 5
step is 10
smaller step is 5


[0, 1, 4, 2, 5, 0, 5]

## 845. Longest Mountain in Array

Let's call any (continguous) subarray B (of A) a mountain if the following properties hold:

* B.length >= 3
* There exists some 0 < i < B.length -1 such that B[0] < B[1] < ... < B[i-1] < B[i] > B[i+1] > ... > B[B.length -1]. 

Given an array A of intergers, return the length of the longest mountain. 
Return 0 if there is no mountain. 

**Example 1:**
Input : [2,1,4,7,3,2,5]

Output: 5 

Explanation: The largest mountain is [1,4,7,3,2] which has length 5. 

**Example 2:**
Input: [2,2,2]

Output: 0

Explanation: There is no mountain. 

In [88]:
def largestMountain(A: str) -> int:
    
    n = len(A)
    
    if n <= 1:
        return 0 
    
    inc = [0]*(n+1)
    dec = [0]*(n+1)
    
    for i in range(1,n):
        if A[i] > A[i-1]:
            inc[i] = inc[i-1] + 1
    for i in range(n-2,-1,-1):
        if A[i] > A[i+1]:
            dec[i] = dec[i+1] + 1 
            
    print(inc)
    print(dec)
            
    ans = 0
    
    for i in range(1,n+1):
        if (inc[i] and dec[i]):
            ans = max(ans, inc[i] + dec[i] + 1)
    return(ans)
            
    
    

In [89]:
A = [2,1,4,7,3,2,5]

largestMountain(A)

[0, 0, 1, 2, 0, 0, 1, 0]
[1, 0, 0, 2, 1, 0, 0, 0]


5

In [90]:
A = [2,2,2]

largestMountain(A)

[0, 0, 0, 0]
[0, 0, 0, 0]


0

## 801.Minimum Swaps To Make Sequences Increasing 


We have two integer sequences A and B of the same non-zero length. 

We are allowed to swap elements A[i] and B[i]. Note both elements are in the same index position in their respective sequences. 

At the end of some number of swaps. A and B are both strictly increasing.

Given A and B, return the minimum number of swaps to make both sequences strictly increasing. It is guaranteed that the given input always makes it possible. 

**Example:**
Input: A = [1,3,5,4], B = [1,2,3,7]

Output: 1

Explanation: 
Swap A[3] and B[3]. Then the sequences are: A = [1,3,5,7] and B = [1,2,3,4] which are both strictly increasing. 

In [2]:
dp = [ [0]*2 for _ in range(5) ]

In [3]:
print(dp)

[[0, 0], [0, 0], [0, 0], [0, 0], [0, 0]]


In [4]:
dp[0][0] = 1

In [6]:
print(dp[1])

[0, 0]


In [113]:
a = float('Inf')

In [114]:
min(a,1)

1

In [121]:
def minSwap(A: list, B: list) -> int:
    
    n = len(A)
    
    if n <= 1:
        return 0 
    
    dp = [[float('Inf')]*2 for _ in range(n) ] 
    dp[0][0] = 0 
    dp[0][1] = 1 
    
    for i in range(1, n):
        if A[i] > A[i-1] and B[i] > B[i-1]:
            dp[i][0] =  dp[i-1][0]
            dp[i][1] =  dp[i-1][1] + 1
        print('i is '+ str(i) + 'after 1st count')
        print(dp)
        if B[i] > A[i-1] and A[i] > B[i-1]:
            dp[i][0] = min(dp[i][0], dp[i-1][1])
            dp[i][1] = min(dp[i][1], dp[i-1][0] + 1)
        print('i is '+ str(i) + 'after 2nd count')
        print(dp)
    return(min(dp[n-1][0], dp[n-1][1])) 

In [122]:
A = [1,3,5,4]
B = [1,2,3,7]

In [123]:
minSwap(A,B)

i is 1after 1st count
[[0, 1], [0, 2], [inf, inf], [inf, inf]]
i is 1after 2nd count
[[0, 1], [0, 1], [inf, inf], [inf, inf]]
i is 2after 1st count
[[0, 1], [0, 1], [0, 2], [inf, inf]]
i is 2after 2nd count
[[0, 1], [0, 1], [0, 2], [inf, inf]]
i is 3after 1st count
[[0, 1], [0, 1], [0, 2], [inf, inf]]
i is 3after 2nd count
[[0, 1], [0, 1], [0, 2], [2, 1]]


1

In [124]:
A = [0,3,5,8,9]
B = [2,1,4,6,9]
minSwap(A,B)

i is 1after 1st count
[[0, 1], [inf, inf], [inf, inf], [inf, inf], [inf, inf]]
i is 1after 2nd count
[[0, 1], [1, 1], [inf, inf], [inf, inf], [inf, inf]]
i is 2after 1st count
[[0, 1], [1, 1], [1, 2], [inf, inf], [inf, inf]]
i is 2after 2nd count
[[0, 1], [1, 1], [1, 2], [inf, inf], [inf, inf]]
i is 3after 1st count
[[0, 1], [1, 1], [1, 2], [1, 3], [inf, inf]]
i is 3after 2nd count
[[0, 1], [1, 1], [1, 2], [1, 2], [inf, inf]]
i is 4after 1st count
[[0, 1], [1, 1], [1, 2], [1, 2], [1, 3]]
i is 4after 2nd count
[[0, 1], [1, 1], [1, 2], [1, 2], [1, 2]]


1

In [125]:
A = [2,4,5,7,10]
B = [1,3,4,5,9]
minSwap(A,B)

i is 1after 1st count
[[0, 1], [0, 2], [inf, inf], [inf, inf], [inf, inf]]
i is 1after 2nd count
[[0, 1], [0, 1], [inf, inf], [inf, inf], [inf, inf]]
i is 2after 1st count
[[0, 1], [0, 1], [0, 2], [inf, inf], [inf, inf]]
i is 2after 2nd count
[[0, 1], [0, 1], [0, 2], [inf, inf], [inf, inf]]
i is 3after 1st count
[[0, 1], [0, 1], [0, 2], [0, 3], [inf, inf]]
i is 3after 2nd count
[[0, 1], [0, 1], [0, 2], [0, 3], [inf, inf]]
i is 4after 1st count
[[0, 1], [0, 1], [0, 2], [0, 3], [0, 4]]
i is 4after 2nd count
[[0, 1], [0, 1], [0, 2], [0, 3], [0, 1]]


0

## Hashed Ports 

Packets are sent to different ports on a compute system based on the hash of their packet ID. The value of the hash is given as below:

Hash = mod(packet_id, numberOfPorts)

where mod is the modulus operator and takes the mod of first operand by second operand. 

The ports are numbered from 0 to (number of ports)-1, and a packet is initially sent to the port that has the port number equal to the hash of its packet ID. Each port requires a time t to send a packet. If a port is currently sending a packet, this packet is then sent to the next port number, and so on. Given that x packets arrive 1 per second, and given the IDs of the packets, find the port at which each packet is finally sent. Firse packet is sent at time t = 1. 

**Function Description:**
Compute the sentTimes function in the editor below. The function must return an integer array denoting the ports at which the packets are sent. 
sentTimes has the following parameters:
* numberOfPorts: An integer, the number of ports in the system
* transmissionTime: An integer, the time for a port to send a packet. 
* packetIds: An integer array, where packetids describes the IDs of the packets in the order in whcy they arrive 

**Constraints:**
* 1 <= numberOfPorts <= 2000
* 1 <= transmissionTime <= 100
* 1 <= x <= 2000
* 1 <= packetIds_i <= 10^5. 


In [10]:
def sentTimes(numberOfPorts, transmissionTime, PacketIds):
    res = []
    t = 1
    queue = []
    avail = [0]*numberOfPorts
    for pid in PacketIds:
        while queue and avail[queue[0]] <= t:
            queue = queue[1:]
        if len(queue) == numberOfPorts:
            t = avail[queue[0]]
            queue = queue[1:]
        port = pid%numberOfPorts
        while avail[port] > t:
            port = (port+1)%numberOfPorts
        avail[port] = t + transmissionTime
        queue.append(port)
        res.append(port)
        t += 1
    return res 

In [4]:
print(sentTimes(numberOfPorts = 5, transmissionTime = 10, PacketIds=[1,2,3,4,5,6]))

[1, 2, 3, 4, 0, 1]


In [6]:
import collections
def hashedPorts(n, T, packet_id):
    ans, t = [], 1
    avail = [0]*n
    queue = collections.deque()
    for pid in packet_id:
        # pop from queue if the port is vailable
        while queue and avail[queue[0]] <= t:
            queue.popleft()
        # If no ports available, wait until there is one available
        if len(queue) == n:
            t = avail[queue.popleft()]
        # try port until find one available
        port = pid%n
        while avail[port] > t:
            port = (port + 1)%n
        # send packet, update available time for the port 
        avail[port] = t+ T
        queue.append(port)
        ans.append(port)
        t += 1
    return(ans)
        

In [11]:
n = 5 
T= 10 
packid = [1,2,3,4,5,6]

print(hashedPorts(n, T, packid))
print(sentTimes(n, T, packid))

[1, 2, 3, 4, 0, 1]
[1, 2, 3, 4, 0, 1]


In [12]:
n = 5
T = 2
packid = [5,1,6,2,7,3]

print(hashedPorts(n, T, packid))
print(sentTimes(n, T, packid))

[0, 1, 2, 3, 2, 3]
[0, 1, 2, 3, 2, 3]


In [13]:
n = 5 
T = 10
packid = [1,1,1,6,2,2]
print(hashedPorts(n,T,packid))
print(sentTimes(n, T, packid))

[1, 2, 3, 4, 0, 1]
[1, 2, 3, 4, 0, 1]


## Multiples of 3 and 5

If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3,5,6,9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below N. 

Input Format.

First line contains T that denotes the number of test cases. This is followed by T lines, each containing an integer N. 

Output Format:

For each test case, print an integer that denotes the sum of all the multiples of 3 or 5 below N. 

In [10]:
8 // 5

1

In [7]:
def sum_multiples_3_5(N):
    
    ans = 0
    
    if N <= 2:
        return ans  
    
    for i in range(3,N):
        if i % 5 == 0 or i % 3 == 0:
            print(i)
            ans += i
            
    return(ans)

In [8]:
N = 10

sum_multiples_3_5(N)

3
5
6
9


23

In [9]:
N = 100

sum_multiples_3_5(N)

3
5
6
9
10
12
15
18
20
21
24
25
27
30
33
35
36
39
40
42
45
48
50
51
54
55
57
60
63
65
66
69
70
72
75
78
80
81
84
85
87
90
93
95
96
99


2318

Time complexity: O(N). Space complexity: O(1). 

In [31]:
def sum_multiples_3_5_v2(N):
    
    ans = 0
    
    for i in range(1,(N-1) //5+1):
        print(i*5)
        ans += (i)*5 
    
    for j in range(1,(N-1)//3 + 1):
        print(j*3)
        ans += (j)*3
        
    for k in range(1,(N-1)//15+1):
        print(k*15)
        ans -= (k)*15
    return(ans)
    

In [32]:
sum_multiples_3_5_v2(100)

5
10
15
20
25
30
35
40
45
50
55
60
65
70
75
80
85
90
95
3
6
9
12
15
18
21
24
27
30
33
36
39
42
45
48
51
54
57
60
63
66
69
72
75
78
81
84
87
90
93
96
99
15
30
45
60
75
90


2318

In [35]:
def sum_multiple_3_5_v3(n):
    
    m = 3 
    x = 5 
    lcm = 15 
    mulm = (n-1) // m 
    mulx = (n -1) // x
    mullcm = (n-1) // lcm
    print(m * mulm * (mulm + 1) //2 + x * mulx * (mulx + 1)//2 - 
         lcm* mullcm *(mullcm +1)// 2)

In [36]:
sum_multiple_3_5_v3(100)

2318


## Factors of 3 and 5

A positive integer is defined as ideal if it can be expressed in the form, $3^x \times 5^y$ for some non-negative integer x and y. Ideal integers have only 3 and 5 as prime divisors. 15, 45, 75 are ideal integers but 6, 10, 21 are not.

Find the number of ideal-integers within the given segment [l,r], inclusive, i.e. the integers t such that l <= t <= r. 

For exampel, l = 200, r = 405. The smallest ideal integer in that range is $3^2 \times 5^2 = 225$. Other values are $3^5 \times 5^0 - 243, 3^1 \times 5^3 = 375, 3^4 *5^1 = 405. $ There are 4 ideal integers in the range. 

In [61]:
def ideal_in_range(l: int, r: int) -> int:
    
    count = 0 
    
    def is_ideal(n: int) -> bool:
        if n == 1:
            return True 
        else:
            if n % 3 == 0:
                print('n is multiple of 3 ' + str(n))
                return is_ideal(n // 3)
            elif n % 5 == 0:
                print('n is mutiple of 5 ' + str(n))
                return is_ideal(n // 5)
            else:
                return False 
    
    for i in range(l,r+1):
        if is_ideal(i) == True:
            print(i)
            count += 1
    return count 

In [62]:
ideal_in_range(1,1)

1


1

In [63]:
ideal_in_range(200, 405)

n is mutiple of 5 200
n is mutiple of 5 40
n is multiple of 3 201
n is multiple of 3 204
n is mutiple of 5 205
n is multiple of 3 207
n is multiple of 3 69
n is multiple of 3 210
n is mutiple of 5 70
n is multiple of 3 213
n is mutiple of 5 215
n is multiple of 3 216
n is multiple of 3 72
n is multiple of 3 24
n is multiple of 3 219
n is mutiple of 5 220
n is multiple of 3 222
n is multiple of 3 225
n is multiple of 3 75
n is mutiple of 5 25
n is mutiple of 5 5
225
n is multiple of 3 228
n is mutiple of 5 230
n is multiple of 3 231
n is multiple of 3 234
n is multiple of 3 78
n is mutiple of 5 235
n is multiple of 3 237
n is multiple of 3 240
n is mutiple of 5 80
n is multiple of 3 243
n is multiple of 3 81
n is multiple of 3 27
n is multiple of 3 9
n is multiple of 3 3
243
n is mutiple of 5 245
n is multiple of 3 246
n is multiple of 3 249
n is mutiple of 5 250
n is mutiple of 5 50
n is mutiple of 5 10
n is multiple of 3 252
n is multiple of 3 84
n is multiple of 3 255
n is mutiple of

4

In [57]:
def is_ideal(n: int) -> bool:
    if n == 1:
        return True 
    else:
        if n % 3 == 0:
            print('n is multiple of 3 ' + str(n))
            return is_ideal(n // 3)
        elif n % 5 == 0:
            print('n is mutiple of 5 ' + str(n))
            return is_ideal(n // 5)
        else:
            return False 

In [58]:
is_ideal(15)

n is multiple of 3 15
n is mutiple of 5 5


True

In [59]:
15 // 3

5

In [60]:
5 // 5

1

Time Complexity: O(n); Space Complexity: O(1)

## 1177. Can You Make a Palindrome 

Given a string s, we make queries on substring of s. 

For each query queries[i] = [left, right, k], we may rearrange the substring s[left], ..., s[right], and then choose up to k of them to replace with any lowercase English letter. 

If the substring is possible to be a palindrome string after the operations above, the result of the query is true. Otherwise, the result is false. 

Return an array answer[], where answer[i] is the result of the i-th query queries[i]. 

Note that: Each letter is counted individually for replacement so if for example s[left...right] = 'aaa', and k = 2, we can only replace two of the letters. (Also, note that the initial string s is never modified by any query. )

Example :

Input: s = "abcda", queries = [[3,3,0],[1,2,0],[0,3,1],[0,3,2],[0,4,1]]
Output: [true,false,false,true,true]
Explanation:
queries[0] : substring = "d", is palidrome.
queries[1] : substring = "bc", is not palidrome.
queries[2] : substring = "abcd", is not palidrome after replacing only 1 character.
queries[3] : substring = "abcd", could be changed to "abba" which is palidrome. Also this can be changed to "baab" first rearrange it "bacd" then replace "cd" with "ab".
queries[4] : substring = "abcda", could be changed to "abcba" which is palidrome.

**Solution:** 

1. Re-arrange is OK. We don't have to worry about how characters are sequences in the substring. "aabbc" is palindrome becuase we can arrange it like abcba. 

2. In palindorme,
 * All characters have even number of duplicates or 
 * One character can have an odd number of duplicates. 
 
 For example, aba has an even-numbered of 'a' and odd-numbered 'b'. So it is a palindrome, Another example 'abab' is also palindorme becuase it has even-numbered of 'a' and 'b'. 
 
3. Another trick here is we can replace some character based on the third parameter ([0,3,1]) from the input query. Example, inpu string: "abcda" and substring for the input [0,3,2] would be "abcd". "abcd" is not palindrom becuase it does not satisfy the condition 1 and 2. But the third parameter, whcih is 2, means we can replace/merge with other odd-numbered duplicate characters. Let's break it down. 

a-1 b-1 c-1 d-1

Now if we merge a with c and b with d, it will look like below:

a-2 b-2

Now the substring will be "abab". After re-arranging, it will become "abba'. How it is a palindrome. 

In [66]:
from collections import Counter 

test_str = "GeeksforGeeks"

res = Counter(test_str)

print(res)

for letter in res:
    print(res[letter] % 2)

Counter({'e': 4, 'G': 2, 'k': 2, 's': 2, 'f': 1, 'o': 1, 'r': 1})
0
0
0
0
1
1
1


In [ ]:
res[l]

In [86]:
from collections import Counter 

def canMakePaliQueries(s:str, queries:list) -> list:
    result = []
    for query in queries:
        result.append(can_make_pali(query))
        
    def can_make_pali(query) -> bool:
        left = query[0]
        right = query[1]
        if left == right:
            return True
        k = query[2]
        rep = 0
        freq = Counter(s[left: right+1])
        
        for letter in freq:
            if freq[letter] %2 ==1 :
                rep += 1
        if rep//2 <= k:
            return True
        else:
            return False 
                
                
        
        
            

In [76]:
s[0:3]

'abc'

In [81]:
def can_make_pali(query,s) -> bool:
    left = query[0]
    right = query[1]
    print('left is '+ str(left))
    print('right is ' + str(right))
    if left == right:
        return True
    k = query[2]
    rep = 0
    freq = Counter(s[left: right+1])
    print('freq is ')
    print(freq)

    for letter in freq:
        if freq[letter] %2 ==1 :
            rep += 1
    print('rep is '+ str(rep))
            
    if rep//2 <= k:
        return True
    else:
        return False 
                

In [85]:
s = 'abcda'

left = 0
right = 4 
k = 1

query = [left, right, k]

can_make_pali(query, s)

left is 0
right is 4
freq is 
Counter({'a': 2, 'b': 1, 'c': 1, 'd': 1})
rep is 3


True

In [87]:
def canMakePaliQueries(s, queries):
    ans = []
    for l, r, k in queries:
        ss = s[l:r+1]
        rem = 0
        for letter, n in collections.Counter(ss).items():
            rem += n % 2
        need = rem // 2
        ans.append(need <= k)
    return ans

However, we are counting letters for every substring, which may overlap with previous calculations. This solution gives Time Limit Exceeded. 

Therefore, we can optimize by caching previous results. We can say dp[i] represents the count of all the letters in s[:i]. Then we know that the counts of dp[l] substracted from dp[r+1] will give us the count of all the letters in the substring s[l:r+1]. We could naively store them all in collections.Counter(). 

In [89]:
def canMakePaliQueries_2(s, queries):
    dp = [Counter()]
    for i in range(1, len(s) + 1):
        dp.append(dp[i-1] + Counter(s[i-1]))
    ans = []
    for l, r, k in queries:
        c = dp[r+1] - dp[l]
        need = sum (v%2 for v in c.values()) //2
        ans.append(need <= k)
    return(ans)

However, the overhead is too great. This solution still gives the Time Limit Exceeded. We can simplify by realizing we only care about lowercase letters, 26 in total. We can store our data in an array of size 26 for each substring. 

In [93]:
N = 5 
dp = [[0]*N]*len(s)
print(dp)

[[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]


In [94]:
s = 'abcda'

i = 2

In [95]:
new = dp[i-1][:]

In [96]:
new

[0, 0, 0, 0, 0]

In [139]:
5 & 1

1

In [149]:
def canMakePaliQueries_3(s, queries):
    N = 26
    a = ord('a')
    dp = [[0]*N]
    for i in range(1, len(s) + 1):
        new = dp[i-1][:]
        j = ord(s[i-1]) - a
        new[j] += 1
        dp.append(new)
        
    ans = []
    
    for l, r, k in queries:
        L = dp[l]
        print(L)
        R = dp[r+1]
        print(R)
        
        ans.append(sum(R[i]- L[i] & 1 for i in range(N)) //2 <= k)
    return ans 

In [150]:
s = 'abcda'

queries = [[3,3,0],[1,2,0],[0,3,1],[0,3,2],[0,4,1]]

In [151]:
canMakePaliQueries_3(s,queries)

[1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[2, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]


[True, False, False, True, True]

## 149. Max Points on a Line

Given n points on a 2D plane, find the maximum number of points that lie on the same straight line. 

**Example 1:**
Input: [[1,1], [2,2], [3,3]]
Output: 3 

**Example 2:**
Input: [[1,1], [3,2], [5,3], [4,1], [2,3], [1,4]]
Output: 4 

In [194]:
from collections import Counter

def maxPoints(points: list) -> int:
    
    def slope(p1,p2):
        x1, y1 = p1
        x2, y2 = p2
        if x1 == x2:
            return 0
        elif y1 == y2:
            return float('Inf')
        else:
            return (x1-x2)/(y1-y2)
    ans = 1
    
    l = len(points)
    if l<= 2:
        return l
        
    for i in range(0, l):
        dup = 0
        count = {}
        p1 = points[i]
        for j in range(i+1, l):
            p2 = points[j]
            if p1 == p2:
                dup += 1
            else:
                s = slope(p1,p2)
                print(s)
                if s in count:
                    count[s] += 1
                else:
                    count[s] = 1 
            print(count)
            if count: 
                ans = max(ans, max(count.values())+ dup + 1)
            else:
                ans = max(ans, dup + 1)
    return ans

Time Complexity: O(n^2)
Space Complexity: O(n)

In [195]:
points = [[1,1], [2,2], [3,3]]
maxPoints(points)

1.0
{1.0: 1}
1.0
{1.0: 2}
1.0
{1.0: 1}


3

In [196]:
points = [[1,1], [3,2], [5,3], [4,1], [2,3], [1,4]]

maxPoints(points)

2.0
{2.0: 1}
2.0
{2.0: 2}
inf
{2.0: 2, inf: 1}
0.5
{2.0: 2, inf: 1, 0.5: 1}
0
{2.0: 2, inf: 1, 0.5: 1, 0: 1}
2.0
{2.0: 1}
-1.0
{2.0: 1, -1.0: 1}
-1.0
{2.0: 1, -1.0: 2}
-1.0
{2.0: 1, -1.0: 3}
0.5
{0.5: 1}
inf
{0.5: 1, inf: 1}
-4.0
{0.5: 1, inf: 1, -4.0: 1}
-1.0
{-1.0: 1}
-1.0
{-1.0: 2}
-1.0
{-1.0: 1}


4

In [180]:
def slope(p1,p2):
    x1 ,y1 = p1
    x2, y2 = p2
    if x1 == x2:
        return 0
    elif y1 == y2:
        return float('Inf')
    else:
        return (x1-x2)/(y1-y2)

In [181]:
points = [[1,1], [2,2], [3,3]]

for i in range(len(points)):
    p1 = points[i]
    print('p1 is '+str(p1[1]))
    for j in range(i+1, len(points)):
        p2 = points[j]
        print('p2 is '+ str(p2))
        s = slope(p1,p2)
        print('slope is ' + str(s))
        

p1 is 1
p2 is [2, 2]
slope is 1.0
p2 is [3, 3]
slope is 1.0
p1 is 2
p2 is [3, 3]
slope is 1.0
p1 is 3


## 1233. Dice Roll Simulation

A die simulator generates a random number from 1 to 6 for each roll. You introduced a constraint to the generator such that it cannot roll the number i more than rollMax[i] (1-indexed) consecutive times. 

Given an array of integers rolMax and an integer n, return the number of distinct sequences that can be obtained with exact n rolls. 

Two sequences are considered different if at least one element differs from each other. Since the answer may be too large, return it modulo 10^9+7. 

**Example 1:**
Input: n = 2, rollMax = [1,1,2,2,2,3]
Output: 34
Explanation: There will be 2 rolls of die, if there are no constraints on the die, there are 6 * 6 = 36 possible combinations. In this case, looking at rollMax array, the numbers of 1 and 2 appear at most once consecutively, therefore sequences (1,1) and (2,2) cannot occur, so the final answer is 36 - 2 = 34. 

**Example 2:**
Input: n = 2, rolMax = [1,1,1,1,1,1]
Output: 30

**Example 3:**
Input: n = 3, rolMax = [1,1,1,2,2,3]
Output: no (1,1), (2,2), (3,3), (4,4,4), (5,5,5)
3 * 3 * 3 - 5 * 2 * 3 - 3 - 2 = 216 - 35 = 181 

In [12]:
from collections import Counter 

rolMax = [1,1,1,2,2,3]

freq = Counter(rolMax)

print(freq)

for item in freq:
    print(freq[item])

Counter({1: 3, 2: 2, 3: 1})
3
2
1


In [15]:
6**3

216

In [21]:
from collections import Counter 
def dieSimulator(n:int, rolMax:list) -> int:
    
    count = 6**n 
    number = 1 
    
    for s in rolMax: 
        print('number is' + str(number))
        print('s is' + str(s))
        if s < n:
            count -= (n-s)*(6**(n-1-s))
            print('count is ' + str(count))
        number += 1
    return (count)


In [26]:
def dieSimulator(n:int, rolMax:list) -> int:
    
    def g(n,k):
        if n < 1:
            return 0
        if n == 1:
            return 1 
        return f(n-1) - f(n- rolMax[k-1]-1) + g(n- rolMax[k-1]-1, k)
    
    def f(n):
        if n < 0:
            return 0
        if n == 0:
            return 1 
        return sum([g(n,k) for k in range(1,7)])
    return f(n) % (10**9 + 7)

Time Complexity: O(n*6)
Space Complexity: O(6*15) 

In [27]:
n = 2
rolMax = [1,1,2,2,2,3]
dieSimulator(n, rolMax)

34

In [28]:
n = 2 
rolMax = [1,1,1,1,1,1]
dieSimulator(n, rolMax)

30

In [29]:
n = 3 
rolMax = [1,1,1,2,2,3]
dieSimulator(n, rolMax)

181

## Turnstile

A university has exactly one turnstile. Unfortunately many people want to pass through the turnstile in different directions. The i-th person comes to turnstile at times[i] and wants to either exit the unversity (direction = 1) or enter the university (direction = 0). People from 2 queues, one to exit one to enter. They are ordered by the time they came to turnstile and if the times are equal by their indices. 

If someone wants to enter the university and other person wants to exit at the same time, there are 3 cases. 

* If in previous second turnstile was not used (maybe it was used before not in the previous second) then the person who wants to leave goes first. 
* If in previous second the turnstile was used to exit then the person who wahts to exit goes first. 
* If in previous second the turnstile was used as an entrance then the person who wants to enter goes first. 

Passint through the turnstile takes 1 second, For each person return the time at whcih they pass throught the turnstile. 

Example: n = 4(there are 4 people). times = [0,0,1,5] (Given in sorted order) it means that person 0 and person 1 come to turnstile at 0th second. direction = [0,1,1,0] (Person 0 wants to enter while person 1 wants to exit). 

Output = [2, 0, 1, 5]

Explanation: -Person 0 and person 1 comes to turnstile at 0th second. Turnstile was not previously used hence person 1 passes through turntile first. Hence ans[1]= 0. 

Now person 0 and person 3 are at turnstile for second 1. But since turnstile was used to exit previously and person 3 wants to exit it will pass though turnstile at 1. Hence ans[2] =1. And so on. 

**Solution:** Use heap, and a global time counter T as "next available time when all people at current second passed". Every time, pop all people whose arrives before T, group by direction. Then according to the rule choose one group to pass and upate T, push the other group back to the heap. Keep on doing this until the heap is empty. 

In [47]:
from heapq import heappush, heappop
def turnstile(n: int, times: list, direction: list) -> list:
    ipq, opq, ans = [],[], [-1]*n
    
    out = True # signal to denote direction of previous second
    
    used = False # signal to denote usedness of previous second
    
    T = 0 # current time 
    
    i = 0
    
    while i < len(times) or ipq or opq:
        
        while i < len(times) and T == times[i]:
            if direction[i] == 1:
                heappush(opq, i)
            if direction[i] == 0:
                heappush(ipq, i)
            i += 1
        print('opq is ')
        print(opq)
        
        print('ipq is ')
        print(ipq)
            
        if ipq or opq:
            
            if not used:
                if opq:
                    ans[heappop(opq)] = T
#                     print(ans)
                    out = True
                else:
                    ans[heappop(ipq)] = T
                    out = False 
                used = True 
            else:
                if out == True and opq:
                    ans[heappop(opq)] = T
                    out = True
                elif ipq:
                    print('T is')
                    print(T)
                    print('used or not')
                    print(used)
#                     print('ipq is ')
#                     print(ipq)
                    ans[heappop(ipq)] = T
                    out = False 
                else:
                    ans[heappop(opq)] = T
                    out = True
                
                    
        else:
            used = False 
        
        
        T += 1
        
    return (ans)
        
        
            
    
            
            
        

In [42]:
n = 4 
times = [0,0,1,5]
direction = [0,1,1,0]

turnstile(n, times, direction)

opq is 
[1]
ipq is 
[0]
opq is 
[2]
ipq is 
[0]
opq is 
[]
ipq is 
[0]
opq is 
[]
ipq is 
[]
opq is 
[]
ipq is 
[]
opq is 
[]
ipq is 
[3]


[2, 0, 1, 5]

In [43]:
n = 7 
times = [0,0,0,0,1,1,3]
direction = [0,0,1,1,1,0,1]

turnstile(n, times, direction)

opq is 
[2, 3]
ipq is 
[0, 1]
opq is 
[3, 4]
ipq is 
[0, 1, 5]
opq is 
[4]
ipq is 
[0, 1, 5]
opq is 
[6]
ipq is 
[0, 1, 5]
opq is 
[]
ipq is 
[0, 1, 5]
opq is 
[]
ipq is 
[1, 5]
opq is 
[]
ipq is 
[5]


[4, 5, 0, 1, 2, 6, 3]

In [48]:
n = 7 
times = [0,0,0,0,1,1,4]
direction = [0,0,1,1,1,0,1]
turnstile(n, times, direction)

opq is 
[2, 3]
ipq is 
[0, 1]
opq is 
[3, 4]
ipq is 
[0, 1, 5]
opq is 
[4]
ipq is 
[0, 1, 5]
opq is 
[]
ipq is 
[0, 1, 5]
T is
3
used or not
True
opq is 
[6]
ipq is 
[1, 5]
T is
4
used or not
True
opq is 
[6]
ipq is 
[5]
T is
5
used or not
True
opq is 
[6]
ipq is 
[]


[3, 4, 0, 1, 2, 5, 6]

Time complexity: O(n)

Space complexity: O(n) for two heap 

## Heap queue (or heapq) in Python 

Heap data structure is mainly used to represent a priority queue. In Python, it is available using "heapq" module. The property of this data structure in python is taht each time the smallest of heap element is popped(min heap). Whenever elements are pushed or popped, heap structure is maintained. The heap[0] element also returns the smallest element each time. 

**Operations on heap:**
1. heapify(iterable):- This function is used to convert the iterable into a heap data structure, i.e., in heap order. 
2. heappush(heap, ele):- This function is used to insert the element mentioned in its arguments into heap. The order is adjusted, so as heap structure is maintained. 
3. heappop(heap):- This functio is used to remove and return the smallest element from heap. The order is adjusted, so as heap structure is maintained. 

In [1]:
import heapq

In [8]:
# initializing list 
li = [5,7,9,1,3]

# using heapify to convert list into heap:
heapq.heapify(li)

# priting created heap
print("the created heap is: ", end="")
print(li)
print(list(li))

the created heap is: [1, 3, 9, 7, 5]
[1, 3, 9, 7, 5]


In [9]:
heapq.heappush(li, 4)
print("The modified heap after push is : ", end="")
print(list(li))

The modified heap after push is : [1, 3, 4, 7, 5, 9]


In [10]:
print("The popped and smallest element is: ", end="")
print(heapq.heappop(li))

The popped and smallest element is: 1


4. heappushpop(heap,ele):- This function combines the functioning of both push and pop operations in one statement, increasing efficiency. Heap order is maintained after this operation. 
5. heapreplace(heap, ele):- This function also inserts and pops element in one statement, but it is different from above function. In this, element is first popped, then element is pushed. i.e., the value larger than the pushed value can be returned. 

In [11]:
li1 = [5,7,9,4,3]
li2 = [5,7,9,4,3]
heapq.heapify(li1)
heapq.heapify(li2)

In [12]:
print("The popped item using heappushpop() is: ", end="")
print(heapq.heappushpop(li1, 2))

The popped item using heappushpop() is: 2


In [13]:
print("The popped item using heapreplace() is: ", end="")
print(heapq.heapreplace(li2,2))

The popped item using heapreplace() is: 3


6. nlargest(k, iterable, key=fun):- This function is used to return the k largest elements from the iterable specified and satisfying the key if mentioned. 

7. nsmallest(k, iterable, key = fun):- This function is used to return the k smallest elements from the iterable specified and satisfying the key if mentioned. 

In [14]:
li1 = [6,7,9,4,3,5,8,10,1]

heapq.heapify(li1)

In [15]:
print("The 3 largest numbers in list are: ", end="")
print(heapq.nlargest(3, li1))

The 3 largest numbers in list are: [10, 9, 8]


In [16]:
print("The 3 smallest numbers in list are: ", end="")
print(heapq.nsmallest(3,li1))

The 3 smallest numbers in list are: [1, 3, 4]


In [27]:
h = []

heapq.heapify(h)

heapq.heappush(h,[0,0,0])

In [28]:
heapq.heappush(h,[0,1,1])

In [29]:
heapq.heappush(h,[1,2,1])

In [30]:
print(h)

[[0, 0, 0], [0, 1, 1], [1, 2, 1]]


In [31]:
print(heapq.heappop(h))

[0, 0, 0]


In [32]:
print(heapq.heappop(h))

[0, 1, 1]


In [34]:
heapq.heappush(h, [5,3,0])

In [35]:
print(heapq.heappop(h))

[1, 2, 1]
